bitwarden_license/src/Commercial.Core/SecretsManager/AuthorizationHandlers/AccessPolicies/ServiceAccountGrantedPoliciesAuthorizationHandler.cs

```diff
@@ -0,0 +1,88 @@
+    private readonly ICurrentContext _currentContext;
+    private readonly IProjectRepository _projectRepository;
+    private readonly IServiceAccountRepository _serviceAccountRepository;
+
+    public ServiceAccountGrantedPoliciesAuthorizationHandler(ICurrentContext currentContext,
+        IAccessClientQuery accessClientQuery,
+        IProjectRepository projectRepository,
+        IServiceAccountRepository serviceAccountRepository)
+    {
+        _currentContext = currentContext;
+        _accessClientQuery = accessClientQuery;
+        _serviceAccountRepository = serviceAccountRepository;
+        _projectRepository = projectRepository;
+    }
+
+    protected override async Task HandleRequirementAsync(AuthorizationHandlerContext context,
+        ServiceAccountGrantedPoliciesOperationRequirement requirement,
+        ServiceAccountGrantedPoliciesUpdates resource)
+    {
+        if (!_currentContext.AccessSecretsManager(resource.OrganizationId))
+        {
+            return;
+        }
+
+        // Only users and admins should be able to manipulate access policies
+        var (accessClient, userId) =
+            await _accessClientQuery.GetAccessClientAsync(context.User, resource.OrganizationId);
+        if (accessClient != AccessClientType.User && accessClient != AccessClientType.NoAccessCheck)
+        {
+            return;
+        }
+
+        switch (requirement)
+        {
+            case not null when requirement == ServiceAccountGrantedPoliciesOperations.Updates:
+                await CanUpdateAsync(context, requirement, resource, accessClient,
+                    userId);
+                break;
+            default:
+                throw new ArgumentException("Unsupported operation requirement type provided.",
+                    nameof(requirement));
+        }
+    }
+
+    private async Task CanUpdateAsync(AuthorizationHandlerContext context,
+        ServiceAccountGrantedPoliciesOperationRequirement requirement, ServiceAccountGrantedPoliciesUpdates resource,
+        AccessClientType accessClient, Guid userId)
+    {
+        var access =
+            await _serviceAccountRepository.AccessToServiceAccountAsync(resource.ServiceAccountId, userId,
+                accessClient);
+        if (access.Write)
+        {
+            var projectIdsToCheck = resource.ProjectGrantedPolicyUpdates.Select(update =>
+                update.AccessPolicy.GrantedProjectId!.Value).ToList();
+
+            var sameOrganization =
+                await _projectRepository.ProjectsAreInOrganization(projectIdsToCheck, resource.OrganizationId);
+            if (!sameOrganization)
+            {
+                return;
+            }
+
+            var projectsAccess =
+                await _projectRepository.AccessToProjectsAsync(projectIdsToCheck, userId, accessClient);
+            if (projectsAccess.Count == projectIdsToCheck.Count && projectsAccess.All(a => a.Value.Write))
```

Comment:
Would it make sense for this logic to happen inside of AccessToProjectsAsync similar to how line 73 sameOrganization returns the boolean we need?

Reply:
In general, I prefer to keep logic out of the repository so it can live in and be unit tested in our queries, commands, and authz handlers.

For this specifically, in the past we had incorporated logic:
 
```
    Task<bool> UserHasReadAccessToProject(Guid id, Guid userId);
    Task<bool> UserHasWriteAccessToProject(Guid id, Guid userId);
``` 

To add support for multiple, it might look like
```
    Task<bool> UserHasReadAccessToAllProjects(List<Guid> ids, Guid userId);
    Task<bool> UserHasReadAccessToSomeProjects(List<Guid>  ids, Guid userId);
    Task<bool> UserHasWriteAccessToAllProjects(List<Guid>  ids, Guid userId);
    Task<bool> UserHasWriteAccessToSomeProjects(List<Guid>  ids, Guid userId);
    //Some other question you might want answered for project access
```
Then, if we add a new permission type, we will have to add more methods to support that.


I prefer only having to maintain two methods:

```
    Task<(bool Read, bool Write)> AccessToProjectAsync(Guid id, Guid userId, AccessClientType accessType);
    Task<Dictionary<Guid, (bool Read, bool Write)>> AccessToProjectsAsync(IEnumerable<Guid> projectIds, Guid userId, AccessClientType accessType);
```
Then it's up to the code calling for the data to determine what to do with it, and it makes it easier to extend in the future.